lib/matplotlib/cm.py

```diff
@@ -371,6 +371,29 @@ def set_array(self, A):
             self._A = None
             return
 
+        if isinstance(self, mpl.collections.QuadMesh):
+            A_ = A.ravel()
+            width, height = self._meshWidth, self._meshHeight
+            if self._shading == 'flat':
+                if not (len(A_) == width * height or
```

Comment:
I think we now need to check all these, not only that they error, but that they work if the set_array is correct?  Like I'm still not sure what happens if you have shading=_flat and pass len(A_) == width*height after the fact.  

Sorry, this might not have really been a good fist issue because of the goofy semantics of this.

Reply:
Ah, no worries :) before any of the changes made, I had a look at this, when len(A) == (width+1)*(height+1) and `shading` = `flat`, the same thing happens where the last dimensions are dropped. When len(A) == width*height, then it's the same thing as setting A with the dimensions of one less than X and Y.
Here, when shading is set to `flat`, regardless of the shape of `C` passed to `pcolormesh`, the height (self._meshHeight) and width (self._meshWidth) are always the same shape as one less than X and Y.
For example : 
```python
# shape of xe is (10,)
# shape of ye is (14,)
# shape of c is (14,10)
col = plt.pcolormesh(xe, ye, c, shading='flat')  # works with drops the dimensions to get (13, 9) accordingly.
# But self.+_meshHeight, self._meshWidth is = (13, 9)

# when 
# shape of xe is (10,)
# shape of ye is (14,)
# shape of c is (13,9)
col = plt.pcolormesh(xe, ye, c, shading='flat')  # works as intended without dropping any dimensions.
# And self.+_meshHeight, self._meshWidth is = (13, 9), same as that when c is of (14, 9)
```
So my point is regardless of the shape of A passed to `set_array`, in the example listed above it should either be (13, 9) or (14, 10) as self._meshHeight and self._meshWidth is already automatically set to (13, 9) if the dimensions are dropped

Reply:
OK, just checking.  I'd forgotten if the dropped row/column happened when it was ingested or when it was drawn...